### PR TITLE
PostgreSQL Predicate support

### PR DESCRIPTION
**Pull request by kirkeaton**
1. Added IndexedPreparedStatement class which maintains an index which is used to set the values of a prepared statements
2. Modified PostgresDAO to use IndexPreparedStatement
3. Added SQLStatement interface
4. Made PropertyInfo, Comparator, and Predicate implement SQLStatement
5. Added createStatement and prepareStatement functions for Mlangs

TODO: test the creation of the SQL queries to validate their correctness

### CODE REVIEW

File: src/foam/core/AbstractPropertyInfo.java [hunk 1]
```diff
@@ -48,12 +49,11 @@ public void toCSV(foam.lib.csv.Outputter outputter, Object value) {
 
   @Override
   public String createStatement() {
-    return getName();
+    return getName().toLowerCase();
```
Comment (kgrgreer) [inline src/foam/core/AbstractPropertyInfo.java:52]: Why is this needed? Isn't Postgres case-insensitive?

File: src/foam/mlang/MLang.java [hunk 2]
```diff
@@ -21,7 +21,9 @@
   public static final Predicate FALSE = new False();
 
   public static Expr prepare(Object o) {
-    return o instanceof Expr ? (Expr) o : new Constant(o);
+    return o instanceof Expr ? (Expr) o :
```
Comment (kgrgreer) [inline src/foam/mlang/MLang.java:24]: Good.

File: src/foam/core/AbstractPropertyInfo.java (hunk 1 above)
Comment (kirkeaton) [reply src/foam/core/AbstractPropertyInfo.java:52]: PostgreSQL converts identifiers to lowercase if they are not double-quoted. So instead of double quoting all table names & column names I figured it would be easier to just convert to lowercase instead.

https://stackoverflow.com/a/20880247

Comment (kgrgreer) [reply src/foam/core/AbstractPropertyInfo.java:52]: Okay, then can we not include this method in the AbtractPropertyInfo and have a custom implementation generated for each one specificiall? This will let us convert to lowercase at compile time and so we'll be faster and generate less garbage. Can be done in future PR.
